src/intelligence_layer/connectors/retrievers/hybrid_qdrant_in_memory_retriever.py

```diff
@@ -0,0 +1,206 @@
+
+
+class HybridQdrantInMemoryRetriever(QdrantInMemoryRetriever):
+    """Search through documents stored in memory using hybrid (keyword + semantic) search.
+
+    This retriever uses a [Qdrant](https://github.com/qdrant/qdrant)-in-Memory vector store instance to store documents and their asymmetric embeddings.
+    When run, the given query is embedded using both a dense and sparse embedding model and scored against the documents in the collection to find the most relevant documents.
+    Finally, the retrievals are fused using the Reciprocal Rank Fusion algorithm.
+
+    Args:
+        documents: The sequence of documents to be made searchable.
+        k: The (top) number of documents to be returned by search.
+        client: Aleph Alpha client instance for running model related API calls.
+        threshold: The minimum value of the fusion rank score (combined cosine similarity and keyword similarity).
+        retriever_type: The type of retriever to be instantiated.
+            Should be `ASYMMETRIC` for most query-document retrieveal use cases, `SYMMETRIC` is optimized
+            for similar document retrieval.
+        distance_metric: The distance metric to be used for vector comparison.
+        sparse_model_name: The name of the sparse embedding model from `fastemebed` to be used.
+
+    Example:
+        >>> from intelligence_layer.connectors import LimitedConcurrencyClient, Document, HybridQdrantInMemoryRetriever
+        >>> client = LimitedConcurrencyClient.from_env()
+        >>> documents = [Document(text=t) for t in ["I do not like rain.", "Summer is warm.", "We are so back."]]
+        >>> retriever = HybridQdrantInMemoryRetriever(documents, 5, client=client)
+        >>> query = "Do you like summer?"
+        >>> documents = retriever.get_relevant_documents_with_scores(query)
+    """
+
+    MAX_WORKERS = 10
+
+    def __init__(
+        self,
+        documents: Sequence[Document],
+        k: int,
+        client: AlephAlphaClientProtocol | None = None,
+        threshold: float = 0.0,
+        retriever_type: RetrieverType = RetrieverType.ASYMMETRIC,
+        distance_metric: Distance = Distance.COSINE,
+        sparse_model_name: str = "bm25",
+    ) -> None:
+        self._client = client or LimitedConcurrencyClient.from_env()
+        self._search_client = QdrantClient(":memory:")
+        self._collection_name = "in_memory_collection"
+        self._k = k
+        self._threshold = threshold
+        self._query_representation, self._document_representation = retriever_type.value
+        self._distance_metric = distance_metric
+
+        self._search_client.set_sparse_model(sparse_model_name)
+        self._sparse_vector_field_name = "text-sparse"
+        self._dense_vector_field_name = "text-dense"
+
+        if self._search_client.collection_exists(collection_name=self._collection_name):
+            self._search_client.delete_collection(
+                collection_name=self._collection_name,
+            )
```

Comment:
What is the reason we delete the collection here? I think this is kind of unintuitive and makes it so i cannot create myself a new retriever when its already running somewhere without affecting it right?

Reply:
edit: I see the existing `QdrantInMemoryRetriever` does the same. Still not the biggest fan but i guess its consistent. What's the reason for not using super init here?

Reply:
Good catch, I have also noted this in the updated PR description. The reason for not using super init is because of a crucial difference in initializing the collections, as now this needs to be done for both the sparse and dense embeddings.

Reply:
I understand your point, but even with this approach, I'd still need to override the default __init__ method because it calls `_add_texts_to_memory`, which is where the logic for storing dense and sparse vectors is defined. Qdrant, however, requires both types of vectors to be stored together for each example when using hybrid search, which is why I chose the current implementation approach (separate `__init__`).